pkg/conversion/cel_converter.go

```diff
@@ -0,0 +1,214 @@
+		if err := apiextensionsv1.Convert_v1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(v.Schema.OpenAPIV3Schema, internalJSONSchemaProps, nil); err != nil {
+			return nil, fmt.Errorf("failed converting version %s validation to internal version: %w", v.Name, err)
+		}
+
+		structuralSchema, err := structuralschema.NewStructural(internalJSONSchemaProps)
+		if err != nil {
+			return nil, fmt.Errorf("error getting structural schema for version %s: %w", v.Name, err)
+		}
+
+		// Deep copy because PruneDefaults is mutating
+		structuralSchema = structuralSchema.DeepCopy()
+
+		if err := structuraldefaulting.PruneDefaults(structuralSchema); err != nil {
+			return nil, fmt.Errorf("error pruning defaults for version %s: %w", v.Name, err)
+		}
+
+		structuralSchemas[v.Name] = structuralSchema
+	}
+
+	compiledRules, err := CompileConversions(apiConversion, structuralSchemas)
+	if err != nil {
+		return nil, fmt.Errorf("error compiling conversion rules: %w", err)
+	}
+
+	return &celConverter{
+		crd:               crd,
+		compiledRules:     compiledRules,
+		structuralSchemas: structuralSchemas,
+		apiConversion:     apiConversion,
+	}, nil
+}
+
+type celConverter struct {
+	crd               *apiextensionsv1.CustomResourceDefinition
+	compiledRules     map[string][]*CompiledRule
+	structuralSchemas map[string]*structuralschema.Structural
+	apiConversion     *apisv1alpha1.APIConversion
+}
+
+func (c *celConverter) Convert(list *unstructured.UnstructuredList, targetGV schema.GroupVersion) (*unstructured.UnstructuredList, error) {
+	convertedList := &unstructured.UnstructuredList{}
+	for i := range list.Items {
+		original := &list.Items[i]
+		converted := original.DeepCopy()
+		converted.SetAPIVersion(targetGV.String())
+
+		originalVersion := original.GetObjectKind().GroupVersionKind().Version
+		compiledRules := c.compiledRules[originalVersion]
+		for _, rule := range compiledRules {
+			if rule.Err != nil {
+				// TODO(ncdc): what to do here?
+				continue
+			}
+
+			fromPath := rule.FromPath
+			if fromPath[0] == '.' {
+				fromPath = fromPath[1:]
+			}
+			fromFields := strings.Split(fromPath, ".")
+			fromValue, exists, err := unstructured.NestedFieldNoCopy(original.Object, fromFields...)
+			if err != nil {
+				return nil, fmt.Errorf("error getting source field %q: %w", rule.FromPath, err)
+			}
+			if !exists {
+				continue
+			}
+
+			toPath := rule.ToPath
+			if toPath[0] == '.' {
+				toPath = toPath[1:]
+			}
+			fields := strings.Split(toPath, ".")
+			if err := unstructured.SetNestedField(converted.Object, fromValue, fields...); err != nil {
+				return nil, fmt.Errorf("error setting target field %q: %w", rule.ToPath, err)
+			}
+
+			if rule.Program == nil {
+				continue
+			}
+
+			// structuralSchema := c.structuralSchemas[originalVersion]
+			bindings := map[string]interface{}{
+				"self": fromValue, // crdcel.UnstructuredToVal(original.Object, structuralSchema),
+			}
```

Comment:
nice, super simple. So we can easily add more values to the context per activation.